### PR TITLE
refactor: switch code base to typescript 

### PR DESCRIPTION
**Pull request by machour**
Well, this was fun :sweat_smile: 

Switched all the code base to typescript, and added the (hopefully) correct types everywhere:
  - Introduced `src/typings` for global interfaces
  - Props & State typings are in the components files, below the component, along with sometime needed global overrides (window, Date)

### CODE REVIEW

File: src/chat/botman.ts [hunk 1]
```diff
@@ -1,31 +1,35 @@
...
     	let data = new FormData();
-    	const postData = {
+    	const postData: { [index: string] : string|Blob } = {
     		driver: 'web',
     		userId: this.userId,
     		message: text,
-    		attachment,
-    		interactive
+    		attachment: attachment as Blob,
+    		interactive: interactive ? '0' : '1' // FIXME:
```
Comment (machour) [inline src/chat/botman.ts]: FormData.append() expects either a String or a Blob as parameter, so I manually converted the boolean to a string.
You may want to check if I didn't break it.

File: src/chat/botman.ts [hunk 2]
```diff
@@ -1,31 +1,35 @@
...
     	};
 
     	Object.keys(postData).forEach(key => data.append(key, postData[key]));
 
     	axios.post(this.chatServer, data).then(response => {
     		const messages = response.data.messages || [];
 
-    		messages.forEach(msg => {
+    		messages.forEach((msg: IMessage) => {
     			if (perMessageCallback) {
```
Comment (machour) [inline src/chat/botman.ts]: _unrelated: move the if before the forEach?_

File: src/widget/configuration.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    desktopWidth: 370,
+    mobileHeight: '100%',
+    mobileWidth: '300px',
+    videoHeight: 160,
+    aboutLink: 'https://botman.io',
+    aboutText: '⚡ Powered by BotMan',
+    chatId: '',
+    userId: '',
+    // FIXME: this is overrided in widget.tsx, probably shouldn't be in conf
+    wrapperHeight: 100,
```
Comment (machour) [inline src/widget/configuration.ts]: :warning: check FIXME

File: src/widget/index.tsx [hunk 4]
```diff
@@ -40,3 +40,7 @@ function injectChat() {
     );
 
 }
+
+declare global {
+    interface Window { attachEvent: Function, botmanWidget: Widget }
+}
```
Comment (machour) [inline src/widget/index.tsx]: I wasn't able to do this globally :/ I'm using something similar in another component as a result

File: src/widget/widget.tsx [hunk 5]
```diff
@@ -136,3 +142,25 @@ export default class Widget extends Component {
...
+}
+
+declare global {
+    interface Window { attachEvent: Function, botmanChatWidget: Api }
+}
+
+// FIXME: toGMTString is deprecated
+interface IDate extends Date {
+  toGMTString?(): string;
+}
```
Comment (machour) [inline src/widget/widget.tsx]: Should we consider switching to toUTCString instead?

File: src/chat/messages/action.tsx [hunk 6]
```diff
@@ -1,12 +1,13 @@
 import { h, Component } from 'preact';
 import {botman} from './../botman';
+import { IMessageTypeProps, IAction, IMessage } from '../../typings';
 
-export default class ActionType extends Component {
+export default class ActionType extends Component<IMessageTypeProps, any> {
```
Comment (machour) [inline src/chat/messages/action.tsx]: IMessageTypeProps is used for all messages types (text, button, ..)

**Comment by mpociot** [comment]
Wow - thank you so much! I looked through the various parts that needed adjustments and fixed them.
Awesome job!

**Pull request merged**